### PR TITLE
metrics: support pulling container images from different registries

### LINKED ISSUE #3968: tests: ci  jenkins-metrics-ubuntu-20-04 test script toomanyrequests rate exceeded
# Description of problem

These was a probability problem of the ci jenkins-metrics-ubuntu-20-04 test script, when build the local-blogbench image.

```
16:08:55 ===== starting test [blogbench] =====
16:08:55 command: docker: yes
16:08:55 command: ctr: yes
16:08:56 command: awk: yes
16:08:56 command: docker: yes
16:08:56 time="2021-09-14T08:08:56Z" level=warning msg="docker.io/library/local-blogbench:latest: image not found"
16:08:56 docker building local-blogbench
16:08:56 Sending build context to Docker daemon   2.56kB

16:08:56 Step 1/7 : FROM public.ecr.aws/lts/ubuntu
16:08:57 latest: Pulling from lts/ubuntu
16:08:58 toomanyrequests: Rate exceeded
16:08:58 ERROR: Failed to docker build image local-blogbench
```

# Expected result

Expect jenkins-metrics-ubuntu-20-04 test running success.

# Actual result

the local-blogbench image build failed, caused the jenkins-metrics-ubuntu-20-04 test running FAILED.

### PR DESCRIPTION
**Pull request by devimc**
Mitigate image registries rate limit issue.
Define a list of public container registries and use it to build
the images used to run metrics.

fixes #3968

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

**Review by GabyCT — APPROVED** [review]

**Review by Jakob-Naucke — COMMENTED** [review]
Hi @devimc -- generally, I really like this solution, it's a lot more hands-off than what we've been doing so far.
DYT this logic could be moved from `metrics` to `.ci` so that other Dockerfiles in the tests repo can use it?

I am also trying to think of ways to improve this in the main repo (osbuilder, static, etc.), but this approach is probably not ideal (reproducible builds, while we don't even have them as of now, become even harder).

**Comment by GabyCT** [comment]
We got an error like
```
18:01:09 INFO: Running qemu metrics tests
18:01:09 /home/jenkins/workspace/workspace/kata-containers-2.0-tests-metrics-ubuntu-20-04-PR/go/src/github.com/kata-containers/tests/.ci/../metrics/lib/common.bash: line 33: `registries[ubuntu]': not a valid identifier
18:01:09 /home/jenkins/workspace/workspace/kata-containers-2.0-tests-metrics-ubuntu-20-04-PR/go/src/github.com/kata-containers/tests/.ci/../metrics/lib/common.bash: line 33: docker.io/library
18:01:09 public.ecr.aws/ubuntu
18:01:09 quay.io/libpod: No such file or directory
18:01:09 [run_metrics_PR_ci.sh:33] ERROR: registries[ubuntu]= docker.io/library
18:01:09 public.ecr.aws/ubuntu
18:01:09 quay.io/libpod
```

File: metrics/lib/common.bash [hunk 1]
```diff
@@ -29,6 +29,12 @@ KSM_PAGES_SHARED="${KSM_BASE}/pages_shared"
 KSM_AGGRESIVE_PAGES=1000
 KSM_AGGRESIVE_SLEEP=50
 
+declare -A registries
+registries[ubuntu]= \
```
Comment (Jakob-Naucke) [inline metrics/lib/common.bash:33]: > ```
> /home/jenkins/workspace/workspace/kata-containers-2.0-tests-metrics-ubuntu-20-04-PR/go/src/github.com/kata-containers/tests/.ci/../metrics/lib/common.bash: line 33: `registries[ubuntu]': not a valid identifier
> /home/jenkins/workspace/workspace/kata-containers-2.0-tests-metrics-ubuntu-20-04-PR/go/src/github.com/kata-containers/tests/.ci/../metrics/lib/common.bash: line 33: docker.io/library
> ```

this should be

```suggestion
registries[ubuntu]=\
```

File: metrics/lib/common.bash [hunk 2]
```diff
@@ -29,6 +29,12 @@ KSM_PAGES_SHARED="${KSM_BASE}/pages_shared"
...
+"docker.io/library
+public.ecr.aws/ubuntu
```
Comment (GabyCT) [inline metrics/lib/common.bash:35]: @devimc I think we need to remove `ubuntu` from here

Comment (GabyCT) [reply metrics/lib/common.bash:35]: @devimc the name of the image should be `public.ecr.aws/lts/ubuntu`

**Comment by GabyCT** [comment]
> Hi @devimc -- generally, I really like this solution, it's a lot more hands-off than what we've been doing so far. DYT this logic could be moved from `metrics` to `.ci` so that other Dockerfiles in the tests repo can use it?
> 
> I am also trying to think of ways to improve this in the main repo (osbuilder, static, etc.), but this approach is probably not ideal (reproducible builds, while we don't even have them as of now, become even harder).

I think that is a good idea but maybe that can be a future improvement in another PR

**Comment by GabyCT** [comment]
The error is
```
16:09:21 sed: -e expression #1, char 31: unknown option to `s'
16:09:21 Sending build context to Docker daemon  3.072kB

16:09:21 Error response from daemon: the Dockerfile (Dockerfile) cannot be empty
16:09:21 sed: -e expression #1, char 36: unknown option to `s'
16:09:21 Sending build context to Docker daemon  3.072kB
```

File: metrics/lib/common.bash (hunk 2 above)
Comment (Jakob-Naucke) [reply metrics/lib/common.bash]: @devimc now it's `public.ecr.aws/lts`, but we need `public.ecr.aws/lts/ubuntu`

Comment (devimc) [reply metrics/lib/common.bash]: I don't think so, ptal https://gallery.ecr.aws/lts/ubuntu

**Usage**

```
Docker CLI

Launch this image locally:

docker run -d --name ubuntu -e TZ=UTC public.ecr.aws/lts/ubuntu:20.04_stable
```

Comment (Jakob-Naucke) [reply metrics/lib/common.bash]: @devimc yes, of course. Missed that the `/ubuntu` is still appended. Sorry for the noise.

**Review by Jakob-Naucke — APPROVED** [review]

**Pull request merged**